scvi/module/_mrdeconv.py

```diff
@@ -245,16 +261,20 @@ def loss(
                 0
             )  # 1, p, n_labels, n_latent
             pre_lse = (
-                Normal(mean_vprior, torch.sqrt(var_vprior)).log_prob(gamma).sum(-1)
+                Normal(mean_vprior, torch.sqrt(var_vprior) + 1e-4)
+                .log_prob(gamma)
+                .sum(-1)
             )  # minibatch, p, n_labels
             log_likelihood_prior = torch.logsumexp(pre_lse, 1) - np.log(
                 self.p
             )  # minibatch, n_labels
             neg_log_likelihood_prior = -log_likelihood_prior.sum(1)  # minibatch
             # mean_vprior is of shape n_labels, p, n_latent
 
-        loss = (
-            n_obs * torch.mean(reconst_loss + kl_weight * neg_log_likelihood_prior)
+        loss = n_obs * (
+            torch.mean(
+                reconst_loss + kl_weight * (neg_log_likelihood_prior + v_sparsity_loss)
```

Comment:
sparsity loss shouldn't be affected by kl annealing except if motivated. Also, sometimes it's a local param (if amortized), sometimes it's a global param (if proportions non amortized). Did we make sure it works in both settings?

Reply:
I tried both non-amortized as well as amortized. The issue is the amortized version: Without annealing the sparsity loss leads to cancelling complete cell types at the beginning of training and the gamma is then not learned correctly, so also at latter time points it's not introducing them. There is no justification for kl_weight and I could also introduce another adaptive weight.